### PR TITLE
Improve error message for adding/removing invalid target

### LINKED ISSUE #3318: Improve error message for `rustup target add invalid-target`
### Problem you are trying to solve

Consider this error:
```
; rustup target add wasm
error: toolchain 'nightly-x86_64-unknown-linux-gnu' does not contain component 'rust-std' for target 'wasm'
note: not all platforms have the standard library pre-compiled: https://doc.rust-lang.org/nightly/rustc/platform-support.html
help: consider using `cargo build -Z build-std` instead
```
The pointer to platform-support.html is good and helpful, but the rest of these messages are a bit misleading. The problem is that `wasm` isn't a target, I need to pick a specific version of wasm:
```
; rustc --print=target-list | grep wasm
wasm32-unknown-emscripten
wasm32-unknown-unknown
wasm32-wasi
wasm64-unknown-unknown
```

### Solution you'd like

Have `rustup` check the output of `rustc --print=target-list` for the specified toolchain and give a more helpful error if it doesn't have the target. Maybe something like this:
```
; rustup target add wasm
error: toolchain 'nightly-x86_64-unknown-linux-gnu' does not support target 'wasm'
note: you can see a list of supported targets with `rustc --print=target-list`
note: if you are adding support for a new target to rustc itself, see https://rustc-dev-guide.rust-lang.org/building/new-target.html
```

### Notes

_No response_

### PR DESCRIPTION
**Pull request by 0xPoe**
close https://github.com/rust-lang/rustup/issues/3318

Check the target first.

### CODE REVIEW

File: src/cli/rustup_mode.rs [hunk 1]
```diff
@@ -1377,7 +1377,7 @@ fn get_target(m: &ArgMatches, distributable: &DistributableToolchain<'_>) -> Opt
 fn component_remove(cfg: &Cfg, m: &ArgMatches) -> Result<utils::ExitCode> {
     let toolchain = explicit_or_dir_toolchain(cfg, m)?;
     let distributable = DistributableToolchain::try_from(&toolchain)?;
-    let target = get_target(m, &distributable);
+    let target: Option<TargetTriple> = get_target(m, &distributable);
```
Comment (rbtcollins) [inline src/cli/rustup_mode.rs]: What requires the type annotation?

File: src/errors.rs [hunk 2]
```diff
@@ -114,6 +117,14 @@ pub(crate) enum RustupError {
         component: String,
         suggestion: Option<String>,
     },
+    #[error("toolchain '{}' does not support target '{}'{}", .desc, .target,
+    "\nnote: you can see a list of supported targets with `rustc --print=target-list`\n\
+    note: if you are adding support for a new target to rustc itself, see https://rustc-dev-guide.rust-lang.org/building/new-target.html
+    ".to_string())]
```
Comment (rbtcollins) [inline src/errors.rs]: Using {} to add in a constant string feels weird. Any reason we can't just put it in the main format string? (perhaps using a multi-line string for easier reading)?

**Review by rbtcollins — COMMENTED** [review]
this looks broadly fine

File: src/cli/rustup_mode.rs (hunk 1 above)
Comment (0xPoe) [reply src/cli/rustup_mode.rs]: I guess my RA got some issues. Reverted.

File: src/errors.rs (hunk 2 above)
Comment (0xPoe) [reply src/errors.rs]: Fixed. Thanks! Make sense.

File: src/errors.rs [hunk 3]
```diff
@@ -114,6 +117,30 @@ pub(crate) enum RustupError {
...
+    } else {
+        "".to_string()
+    })]
+    UnknownTarget {
+        desc: ToolchainDesc,
+        target: TargetTriple,
+        suggestion: Option<String>,
+    },
+    #[error("toolchain '{}' does not have target '{}' installed{}\n", .desc, .target,
+    if let Some(suggestion) = .suggestion {
```
Comment (rbtcollins) [inline src/errors.rs:134]: This is duplicated. What about a helper? `suggest_message(.suggestion)` ?

**Review by rbtcollins — APPROVED** [review]

**Comment by rbtcollins** [comment]
I'm merging this, but have left a suggestion to reduce code duplication

**Comment by jyn514** [comment]
this is awesome, thank you both ❤️❤️

**Pull request merged**